src/main/scala/org/ergoplatform/nodeView/ErgoNodeViewHolder.scala

```diff
@@ -132,13 +131,18 @@ abstract class ErgoNodeViewHolder[State <: ErgoState[State]](settings: ErgoSetti
   }
 
 
-  protected def requestDownloads(pi: ProgressInfo[BlockSection]): Unit =
-    pi.toDownload.foreach { case (tid, id) =>
-      context.system.eventStream.publish(DownloadRequest(tid, id))
+  protected def requestDownloads(pi: ProgressInfo[BlockSection]): Unit = {
+    val toDownload = mutable.Map[ModifierTypeId, Seq[ModifierId]]()
+    pi.toDownload.foreach { case (tid, mid) =>
+      toDownload.put(tid, toDownload.getOrElse(tid, Seq()) :+ mid)
```

Comment:
This `:+` operation makes the loop O(n ^ 2)

Reply:
Actually, this map contains 1 modifier id per type at most, see where `ProgressInfo` is constructed, the only branch where  `toDownload` is not empty is `ProgressInfo(None, Seq.empty, toProcess, toDownload(h))` , where h is a single header

Reply:
@kushti The statements above are good to be placed in the code comments. Such code invariants can be useful when refactoring and reading the code.